### PR TITLE
Python: Rename skills to plugins. Update prompt template config to use execution settings to align with dotnet.

### LINKED ISSUE #3319: Python: Skills to plugins

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

To better align with SK dotnet v1, Python skills have been renamed to plugins. This removes the root samples skills directory as dotnet supports the plugins and their configs. This the beginning of more work to consolidate other files/names, per the backlog. I've run several greps for files or file names containing "skill" (case insensitive), and no results appear. Do alert if you find anything else that includes "skill."

Fixes #3319 

### Description

This is a large, breaking change PR as the replacement of "skills" to "plugins" affects many files. It was difficult to break this up into smaller chunks without leaving the SK Python repo in a bad state for some time. Other updates include:

- Prompt template has been updated to use `execution_settings`, which aligns with dotnet, from the previous `completion` that Python used.
- AzureOpenAI/OpenAI function calling has been updated to match the current versions of tools/tool_choice. Parsing messages in the open_ai utils respects this as it now looks for tool_calls and it configures the function_call with the id (new), name, and argument). 
- File renames, where the filename used to contain "skill" will look like new code, but it's the same code, just with "plugin" in the name.
- Kernel examples and notebooks have been tested and are working. Unit tests and integration tests are passing.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
   "[csharp]": {
     "editor.defaultFormatter": "ms-dotnettools.csharp",
     "editor.codeActionsOnSave": {
-      "source.fixAll": true
+      "source.fixAll": "explicit"
```
Comment (moonbox3) [inline .vscode/settings.json:12]: Not sure why this file was updated. Will revert.

File: python/notebooks/03-semantic-function-inline.ipynb [hunk 2]
```diff
@@ -84,7 +84,7 @@
     "    kernel.add_text_completion_service(\"dv\", azure_text_service)\n",
     "else:\n",
     "    api_key, org_id = sk.openai_settings_from_dot_env()\n",
-    "    oai_text_service = OpenAITextCompletion(ai_model_id=\"text-davinci-003\", api_key=api_key, org_id=org_id)\n",
+    "    oai_text_service = OpenAITextCompletion(ai_model_id=\"gpt-4\", api_key=api_key, org_id=org_id)\n",
```
Comment (eavanvalkenburg) [inline python/notebooks/03-semantic-function-inline.ipynb:87]: either use gpt-35-turbo-instruct (a text model) or change the class to OpenAIChatCompletion, probably the first.

File: python/notebooks/05-using-the-planner.ipynb [hunk 3]
```diff
@@ -53,8 +53,8 @@
     "else:\n",
     "    api_key, org_id = sk.openai_settings_from_dot_env()\n",
     "    kernel.add_chat_service(\n",
-    "        \"gpt-3.5\",\n",
-    "        OpenAIChatCompletion(ai_model_id=\"gpt-3.5-turbo\", api_key=api_key, org_id=org_id),\n",
+    "        \"gpt-4\",\n",
+    "        OpenAIChatCompletion(ai_model_id=\"gpt-4\", api_key=api_key, org_id=org_id),\n",
```
Comment (eavanvalkenburg) [inline python/notebooks/05-using-the-planner.ipynb:57]: why was this changed?

File: python/samples/kernel-syntax-examples/plugins_from_dir.py [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import asyncio
+import os
+
+import semantic_kernel as sk
+import semantic_kernel.connectors.ai.open_ai as sk_oai
+
+kernel = sk.Kernel()
+
+useAzureOpenAI = True
+model = "text-davinci-002"
```
Comment (eavanvalkenburg) [inline python/samples/kernel-syntax-examples/plugins_from_dir.py:12]: maybe also update this one

File: python/samples/kernel-syntax-examples/self-critique_rag.py [hunk 5]
```diff
@@ -50,7 +50,7 @@ async def main() -> None:
     kernel.add_text_completion_service(
         "dv",
         AzureTextCompletion(
-            deployment_name="gpt-35-turbo-instruct",
+            deployment_name="text-davinci-003",
```
Comment (eavanvalkenburg) [inline python/samples/kernel-syntax-examples/self-critique_rag.py:53]: text-davinci is depricated

**Review by juliomenendez — CHANGES_REQUESTED** [review]
On the 6, 8, 9, 11 notebooks the SK version is changing to 0.4.4, is that on purpose?
Couple of comments, looks good otherwise.

File: python/semantic_kernel/plugin_definition/parameter_view.py [hunk 6]
```diff
@@ -11,8 +13,8 @@ class ParameterView(SKBaseModel):
     name: str
     description: str
     default_value: str
-    type_: str = Field(default="string", alias="type")
-    required: bool = False
+    type_: Optional[str] = Field(default="string", alias="type")
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/plugin_definition/parameter_view.py:16]: Why this change?

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
first quick look, not bad overall, some notes and questions added!

File: python/samples/kernel-syntax-examples/openai_function_calling.py [hunk 7]
```diff
@@ -25,67 +25,68 @@
...
 )
 prompt_template = sk.ChatPromptTemplate[OpenAIChatMessage](
     "{{$user_input}}", kernel.prompt_template_engine, prompt_config
 )
-prompt_template.add_system_message(system_message)
-prompt_template.add_user_message("Hi there, who are you?")
-prompt_template.add_assistant_message("I am Mosscap, a chat bot. I'm trying to figure out what people need.")
+# prompt_template.add_system_message(system_message)
+# prompt_template.add_user_message("Hi there, who are you?")
+# prompt_template.add_assistant_message("I am Mosscap, a chat bot. I'm trying to figure out what people need.")
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/samples/kernel-syntax-examples/openai_function_calling.py:89]: Good catch. Was testing something, and didn't revert. Thanks.

File: python/samples/kernel-syntax-examples/openai_function_calling.py [hunk 8]
```diff
@@ -97,7 +98,8 @@ async def main() -> None:
     context.variables["user_input"] = "I want to find a hotel in Seattle with free wifi and a pool."
 
     context = await chat_function.invoke_async(context=context)
-    if function_call := context.objects.pop("function_call"):
+    function_call = context.objects.get("function_call", None)  # Default to None if not found
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/samples/kernel-syntax-examples/openai_function_calling.py:101]: Ha. Was testing something, and re-wrote like that. Going to add back in the walrus operator!

File: python/semantic_kernel/connectors/ai/open_ai/utils.py [hunk 9]
```diff
@@ -188,12 +188,15 @@ def _parse_message(
...
-        )
+    tool_calls = message.tool_calls if hasattr(message, "tool_calls") else None
+    function_calls = (
+        [FunctionCall(id=call.id, name=call.function.name, arguments=call.function.arguments) for call in tool_calls]
+        if tool_calls
+        else None
+    )
+
+    # todo: support multiple function calls
+    function_call = function_calls[0]
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/utils.py:199]: Great catch!

File: python/semantic_kernel/semantic_functions/prompt_template_config.py [hunk 10]
```diff
@@ -15,65 +15,67 @@ class PromptTemplateConfig(SKBaseModel, Generic[AIRequestSettingsT]):
...
-                    )
-                if "defaultValue" in parameter:
-                    defaultValue = parameter["defaultValue"]
-                else:
-                    raise Exception(
-                        f"Input parameter '{name}' doesn't have a default value (function: {config['description']})"
-                    )
+        config = cls._process_execution_settings(config, data)
+
+        if "input_variables" in data:
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/semantic_kernel/semantic_functions/prompt_template_config.py:31]: I don't see any failures. In some examples, from the root samples directory, we load the `FunPlugin` which has `Excuses` (no `"input_variables"`) and the `Joke` which does have `input_variables`. While working on this, I realized that no "skills" had their `"input_variables"` loading before. The previous code had a bug, it seems.

File: python/semantic_kernel/plugin_definition/constants.py [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import typing as t
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/semantic_kernel/plugin_definition/constants.py:1]: Looks like we do, as the `GLOBAL_PLUGIN`'s type is specified as `t.Final[str]`. I did add a missing header here, though. :)

File: python/notebooks/03-semantic-function-inline.ipynb (hunk 2 above)
Comment (moonbox3) [reply python/notebooks/03-semantic-function-inline.ipynb:87]: Good catch, thanks. Was testing something, and realized gpt-4 can't be used as a text completion model. :) Forgot to update it.

File: python/notebooks/05-using-the-planner.ipynb (hunk 3 above)
Comment (moonbox3) [reply python/notebooks/05-using-the-planner.ipynb:57]: While running that notebook with gpt-3.5-turbo, it threw an error in the last cell that the number of tokens (both input + completion) exceeded the allowed amount for that model. It seems I just need to set it to `gpt-3.5-turbo-1106` which has a larger context window.

File: python/semantic_kernel/plugin_definition/parameter_view.py (hunk 6 above)
Comment (moonbox3) [reply python/semantic_kernel/plugin_definition/parameter_view.py:16]: As part of the `prompt_template_config` changes I did to support plugins (like those use by dotnet in the root of the SK directory), the `"input_variables"` block was never hit. The `type_` and `required` params appear to be optional in those `"input_variables"`. Updated this here to support that.

**Comment by moonbox3** [comment]
> On the 6, 8, 9, 11 notebooks the SK version is changing to 0.4.4, is that on purpose? Couple of comments, looks good otherwise.

The rename to 0.4.4.dev0 is a mistake. Fixing! Thank you.

File: python/semantic_kernel/plugin_definition/constants.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (juliomenendez) [reply python/semantic_kernel/plugin_definition/constants.py:2]: Yeah, I meant why use alias? Couldn't this file just be:
```python
# Copyright (c) Microsoft. All rights reserved.
from typing import Final

GLOBAL_PLUGIN: Final[str] = "_GLOBAL_FUNCTIONS_"
```

**Review by juliomenendez — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/plugin_definition/constants.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (juliomenendez) [reply python/semantic_kernel/plugin_definition/constants.py:2]: Discussed offline. The alias is required.

**Review by alexchaomander — APPROVED** [review]
Overall looks good to me!

**Review by juliomenendez — APPROVED** [review]
LGTM

File: python/semantic_kernel/semantic_functions/prompt_template_config.py [hunk 12]
```diff
@@ -15,65 +15,67 @@ class PromptTemplateConfig(SKBaseModel, Generic[AIRequestSettingsT]):
     schema_: int = Field(default=1, alias="schema")
     type: str = "completion"
     description: str = ""
-    completion: AIRequestSettingsT = Field(default_factory=AIRequestSettings)
+    execution_settings: AIRequestSettingsT = Field(default_factory=AIRequestSettings)  # todo: this should be a dict
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/semantic_functions/prompt_template_config.py:18]: Let's make a note to also rename the request settings to ExecutionSettings?

**Review by eavanvalkenburg — APPROVED** [review]
LGTM

File: python/semantic_kernel/semantic_functions/prompt_template_config.py (hunk 12 above)
Comment (moonbox3) [reply python/semantic_kernel/semantic_functions/prompt_template_config.py:18]: Good idea, will add a work item. Thanks!

**Review by juliomenendez — APPROVED** [review]
